src/backend/common/dispatch.hpp

```diff
@@ -42,3 +44,73 @@ inline T greatestPrimeFactor(T n) {
 
     return v;
 }
+
+// For OPENCL, the dimensions of local are returned
+// usage: cl::NDRange local = bestBlockSize<cl::NDRange>(dims, WG)
+// For CUDA, the dimensions of 1 block are returned
+// usage: dim3 block = bestBlockSize<dim3>(dims, 32);
+// The parameter dims can have any type as long as it is convertable to unsigned
+
+// Remark: The bestBlockSize is only best for independent element operations, as
+// are: copying, scaling, math on independent elements, ...
+// Since vector dimensions can be returned, it is NOT USABLE FOR BLOCK
+// OPERATIONS, as are: matmul, etc.
+template<typename Tout, typename Tin>
+Tout bestBlockSize(const Tin dims[4], unsigned warp) {
+    const unsigned d0         = static_cast<unsigned>(dims[0]);
+    const unsigned d1         = static_cast<unsigned>(dims[1]);
+    const unsigned d2         = static_cast<unsigned>(dims[2]);
+    const unsigned OCC        = 3;
+    const unsigned elements   = d0 * d1;
+    const unsigned minThreads = warp / 4;  // quarter wave
+    const unsigned maxThreads =
+        std::min(warp * 4, divup(elements * warp, 16384U) * minThreads);
```

Comment:
The idea here is that for very small array dimensions, the opposite is needed (lower occupation).

Suppose a GPU can handle 1024 threads (32*32 warps) in parallel.  Suppose our array is a vector with dim0=256 elements.
- When we go for maximum occupation, we set a block size of 32, resulting in the activation of 8 parallel warps (=8CPUs).  The other 24CPUs will be idle.  We also know that for each warp only 8 threads are really in parallel, so 64 real parallel threads for the whole system with 4 loops each.
- Would it not better to use all the available CPUs (32) with only 1 loop.  This is achieved by launching with a block size of 8 (=minThreads).  The occupation is here only 25%, although 4x faster than with 100%.

As long as all elements of an array is smaller than the available parallel threads, we limit the maxThreads for 1 block to create the occupation inefficiency described above.
The number 16384U is an estimated average (=512*32; 512 stream processors) over all GPU's I looked up.  Again this number does not has to be precise, since the impact decreases the closer we get there.  On top an error margin of 50% is allowed.

Increasing of lowering it did not have much of an impact.  Removing it generated visible slowdowns.
I reused the same number later, because I did not have anything better.  The reasoning is the same, why increase the looping inside a thread when all threads are not occupied.


Do you want me to include all the above comments inside the code?

Reply:
Sorry, but I am confused by the terms here: CPUs, parallel warps. Let's stick to NVIDIA architecture for ease of explanation. Can you please use their jargon: SM, Warp and thread to explain the logic here.

Reply:
I'll try to stick to NVIDIA nomenclature.

Suppose GPU has 4 cores (SP), so 32*4=128 threads can run in parallel.
Suppose our array is an vector of 32 floats, all in dim0.

General rule: 
each SP always receives a full warp.  This is achieved by extending threads up to a multiple of the warp. Empty threads do not consume cycles and are skipped.
When 8 threads execute the same instruction, the SIMD will execute in parallel (1 cycle).
When 7 threads execute the same instruction, the ALU will execute in serial (7 cycles).

- We set threads(128,1,1) and blocks(1,1,1), resulting in
                 ----------------------------------------------blockIdx0----------------------------------------------------
     cycle1: SP0:threadIdx   0..  7; SP1:threadIdx 33..40; SP2:threadIdx 65..72; SP3:threadIdx   97..104
     cycle2: SP0:threadIdx   8..15; SP1:threadIdx 41..48; SP2:threadIdx 73..80; SP3:threadIdx 105..112
     cycle3: SP0:threadIdx 16..23; SP1:threadIdx 49..56; SP2:threadIdx 81..88; SP3:threadIdx 113..120
     cycle4: SP0:threadIdx 24..32; SP1:threadIdx 57..64; SP2:threadIdx 89..96; SP3:threadIdx 121..128
   --> 25% of available SP is used (threads 33..128 exit immediately) and 4 cycles needed.
   --> Occupation = 100%

- We set threads(32,1,1) and blocks(1,1,1), resulting in
                 ------blockIdx0--------    -------------IDLE--------------
     cycle1: SP0:threadIdx   0..  7; SP1:idle; SP2:idle; SP3:idle
     cycle2: SP0:threadIdx   8..15; SP1:idle; SP2:idle; SP3:idle
     cycle3: SP0:threadIdx 16..23; SP1:idle; SP2:idle; SP3:idle
     cycle4: SP0:threadIdx 24..32; SP1:idle; SP2:idle; SP3:idle
   --> 25% of available SP is used and 4 cycles needed. 
   --> Occupation = 25%

- We set threads(8,1,1) and blocks(4,1,1), resulting in
                 ------blockIdx0----- ------blockIdx1------  -------blockIdx2------- -------blockIdx3-------
     cycle1: SP0:threadIdx 0..7; SP1:threadIdx 8..15; SP2:threadIdx 16..23; SP3:threadIdx 24..32
     cycle2, cycle3, cycle4 are skipped since no threads to launch
     --> 100% of available SP is used and 1 cycle needed. 
     --> Occupation = 100%

Assume that we only have 16 floats, all in dim0
- We set threads(4,1,1) and blocks(4,1,1), resulting in
                 ----blockIdx0---- ----blockIdx1---- ----blockIdx2----   ----blockIdx3----
     cycle1: SP0:threadIdx 0; SP1:threadIdx 4; SP2:threadIdx 8; SP3:threadIdx12
     cycle2: SP0:threadIdx 1; SP1:threadIdx 5; SP2:threadIdx 9; SP3:threadIdx13
     cycle3: SP0:threadIdx 2; SP1:threadIdx 6; SP2:threadIdx10; SP3:threadIdx14
     cycle4: SP0:threadIdx 3; SP1:threadIdx 7; SP2:threadIdx11; SP3:threadIdx15
     --> 100% of available SP is used and 4 cycles needed.
     --> Occupation = 100%

- We set threads(8,1,1) and blocks(1,1,1), resulting in
                  -----blockIdx0------
     cycle 1: SP0:threadIdx 0..7; SP1:Idle; SP2:Idle; SP3:Idle
     --> 25% of available SP is used and 1 cycle needed.
     --> Occupation = 25%

Conclusion:
- Occupation rate is useless for very small arrays;  Nr cycles counts!!
- Minimum is warp/4 (// lines of SIMD), because this will 1 cycle == 1 individual instruction
    

I looked up multiple GPU's and they vary between 512 cores (Radeon R7, Vega8) up to 10496 cores (RTX-3090).  As average, I wanted to take 1024 cores (16384 threads on AMD).  I also assumed that NVIDIA users would have bigger cards, so there I took 2048 cores.
In case the user has less cores: All cores will be used and cycles increase (scheduling overhead)
In case the user has more cores: 4x faster compared to nothing (all cores are not used, no impact when memory latency is covered by 16384 threads) 
    AMD: (elements * 64 / 16384) * 16 = elements / (16*16) *16  --> 16 cores used
    NVIDIA: (elements * 32 / 16384) * 8 = elements / (64*8) * 8   --> 64 cores used

I WILL RERUN THE PROFILING, because I think it should be
   AMD : (elements / (1024*16) * 16 = (elements / 16384) * 16 --> divup(d0*d1,1<<14) * minThreads
   NVIDIA: (elements / (2048*8) * 8 = (elements / 16384) * 8    --> divup(d0*d1,1<<14) * minThreads

Reply:
First of all I think this is fantastic work. Your logic here is valid. I think this can be improved if you take the number of compute units into account with your calculations. You can pass it into the function  this using getDeviceProp(id).multiProcessorCount on cuda and `getDevice().getInfo<CL_DEVICE_MAX_COMPUTE_UNITS>()` on OpenCL. This will give you a general idea of the number of compute resources available on the device.

You will need to figure out the number of ALUs available on each compute unit. I think  a safe minimum is 64. This number matches the minimum compute unit for AMD and NVIDIA GPUs. Intel GPUs have 8 shader units per compute unit but lets ignore them for now.

I think that moving below 64 threads per work-group will give you diminishing returns. I think your model of the GPU scheduling is incorrect with respect to the minimum number of clocks required to do an operation. I don't  think you can minimize the impact of an inactive thread in the wavefront like you mention in your response. If you have different results then I would love to see it. I love to be proven wrong :).

Optimizing for very small data sizes will probably not yield favorable results. For these data sizes the bottleneck is going to be either driver overhead, memory loads overhead or other processing overhead internal to ArrayFire. From an engineering point of view, it might be better to simplify the logic to increase maintainability(if you don't see significant gains).

>When 8 threads execute the same instruction, the SIMD will execute in parallel (1 cycle).
When 7 threads execute the same instruction, the ALU will execute in serial (7 cycles).

I am not sure I understand you here. Can you please clarify.

Reply:
As mentioned, I will rerun multiple performance tests with only this parameter changed.
It is possible, that due multiple improvements before on all parameters, that I got wrong conclusions.
More to come.

From what I understand from SIMD, is that all // lines have to be occupied or none at all.  This has as consequence that when only 7 data lines are possible, that the SP no longer can use the SIMD (1x 8//) and is forced to use the ALU (7x 1) instead.  Both do use however the same number of cycles.

Reply:
> I think that moving below 64 threads per work-group will give you diminishing returns. I think your model of the GPU scheduling is incorrect with respect to the minimum number of clocks required to do an operation. I don't think you can minimize the impact of an inactive thread in the wavefront like you mention in your response. If you have different results then I would love to see it. I love to be proven wrong :)

As mentioned in my last comment, I think I am incorrect here. I can see situations where smaller block sizes (<64 threads) can be beneficial for more expensive operations such as doubles and special functions. Please disregard that comment.

> From what I understand from SIMD, is that all // lines have to be occupied or none at all. This has as consequence that when only 7 data lines are possible, that the SP no longer can use the SIMD (1x 8//) and is forced to use the ALU (7x 1) instead. Both do use however the same number of cycles.

I don't think I have heard of that. All the documentation I have read treats the inactive threads as masked off and the operations still use one cycle even though 7 are active. On AMD hardware I understand that they have a scalar execution units which comes into play when all threads perform the same operation on the same data. This is used for control flow and address operations typically.